cmd/cosign/webhook/main.go

```diff
@@ -17,99 +17,78 @@ package main
-		o.Development = true
-	}))
+var secretName = flag.String("secret-name", "", "The name of the secret in the webhook's namespace.")
 
-	var (
-		metricsAddr        = net.ParseIP("127.0.0.1")
-		metricsPort uint16 = 8080
+func main() {
+	opts := webhook.Options{
+		ServiceName: "webhook",
+		Port:        8443,
+		SecretName:  "webhook-certs",
+	}
+	ctx := webhook.WithOptions(signals.NewContext(), opts)
 
-		bindAddr        = net.ParseIP("0.0.0.0")
-		bindPort uint16 = 8443
+	// Allow folks to configure the port the webhook serves on.
+	flag.IntVar(&opts.Port, "secure-port", opts.Port, "The port on which to serve HTTPS.")
 
-		tlsCertDirectory string
-		secretKeyRef     string
+	// This calls flag.Parse()
+	sharedmain.MainWithContext(ctx, "cosigned",
+		certificates.NewController,
+		NewValidatingAdmissionController,
 	)
+}
 
-	klog.InitFlags(goflag.CommandLine)
-	flags := flag.NewFlagSet("main", flag.ExitOnError)
-	flags.AddGoFlagSet(goflag.CommandLine)
-	flags.StringVar(&secretKeyRef, "secret-key-ref", "", "The secret that includes pub/private key pair")
-	flags.IPVar(&metricsAddr, "metrics-address", metricsAddr, "The address the metric endpoint binds to.")
-	flags.Uint16Var(&metricsPort, "metrics-port", metricsPort, "The port the metric endpoint binds to.")
-	flags.IPVar(&bindAddr, "bind-address", bindAddr, ""+
-		"The IP address on which to listen for the --secure-port port.")
-	flags.Uint16Var(&bindPort, "secure-port", bindPort, "The port on which to serve HTTPS.")
-	flags.StringVar(&tlsCertDirectory, "tls-cert-dir", tlsCertDirectory, "The directory where the TLS certs are located.")
-
-	err := flags.Parse(os.Args[1:])
-	if err != nil {
-		klog.Error(err)
-		os.Exit(1)
-	}
-
-	cosignedValidationFuncs := map[schema.GroupVersionKind]webhook.ValidationFunc{
-		corev1.SchemeGroupVersion.WithKind("Pod"):          webhook.ValidateSignedResources,
-		batchv1.SchemeGroupVersion.WithKind("Job"):         webhook.ValidateSignedResources,
-		appsv1.SchemeGroupVersion.WithKind("Deployment"):   webhook.ValidateSignedResources,
-		appsv1.SchemeGroupVersion.WithKind("StatefulSet"):  webhook.ValidateSignedResources,
-		appsv1.SchemeGroupVersion.WithKind("ReplicateSet"): webhook.ValidateSignedResources,
-		appsv1.SchemeGroupVersion.WithKind("DaemonSet"):    webhook.ValidateSignedResources,
-	}
-
-	cosignedValidationHook := webhook.NewFuncAdmissionValidator(webhook.Scheme, cosignedValidationFuncs, secretKeyRef)
+func NewValidatingAdmissionController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	validator := cwebhook.NewValidator(ctx, *secretName)
 
-	opts := ctrl.Options{
-		Scheme:             webhook.Scheme,
-		MetricsBindAddress: net.JoinHostPort(metricsAddr.String(), strconv.Itoa(int(metricsPort))),
-		Host:               bindAddr.String(),
-		Port:               int(bindPort),
-		CertDir:            tlsCertDirectory,
-	}
+	return validation.NewAdmissionController(ctx,
+		// Name of the resource webhook.
+		"cosigned.sigstore.dev",
```

Comment:
nit: consider making this a const with a comment that if this is changed, you also have to update config/200-clusterrole to match this.